src/api/status/src/js/github-stats.js

```diff
@@ -1,74 +1,108 @@
   // get latest author from commits list: https://docs.github.com/en/rest/reference/repos#list-commits
-  try {
-    const res = await fetchGitHubApi(owner, repo, 'commits');
-    const commitResponse = await res.json();
-
-    if (!res.ok) throw new Error(`[Error ${res.status}] - ${commitResponse.message}`);
-
-    const lastCommitResponse = commitResponse[0];
-    returnedValue.avatar = lastCommitResponse.author.avatar_url;
-    returnedValue.username = lastCommitResponse.author.login;
-    returnedValue.fullName = lastCommitResponse.commit.author.name;
-    returnedValue.commitURL = lastCommitResponse.html_url;
-    returnedValue.shortSha = lastCommitResponse.sha.substr(0, 7);
-    returnedValue.usernameURL = `https://github.com/${lastCommitResponse.author.login}`;
-  } catch (error) {
-    console.error(error);
+  const res = await fetchGitHubApi(owner, repo, 'commits');
+  const commitResponse = await res.json();
+
+  if (!res.ok) {
+    throw new Error(`[Code ${res.status}] - ${commitResponse.message}`);
   }
 
+  const lastCommitResponse = commitResponse[0];
+
+  return {
+    avatar: lastCommitResponse.author.avatar_url || '',
+    username: lastCommitResponse.author.login || '',
+    fullName: lastCommitResponse.commit.author.name,
+    commitURL: lastCommitResponse.html_url,
+    shortSha: lastCommitResponse.sha.substr(0, 7),
+    usernameURL: `https://github.com/${lastCommitResponse.author.login || ''}`,
+  };
+};
+
+const getContributorsInfo = async (owner, repo) => {
   // get total contributors: https://docs.github.com/en/rest/reference/repos#list-repository-contributors
+  const res = await fetchGitHubApi(owner, repo, 'contributors?per_page=1');
+
+  if (res.status >= 400) {
+    throw new Error(`[Code ${res.status}] - ${(await res.json()).message}`);
+  }
+
+  const contributorsResponse = await res.headers.get('link');
+  const [, totalContributors] = contributorsResponse.match(/.*"next".*&page=([0-9]*).*"last".*/);
+
+  return {
+    totalContributors,
+  };
+};
+
+module.exports = async function getGitHubData(owner, repo) {
+  let returnGithubInfo = {};
   try {
-    const res = await fetchGitHubApi(owner, repo, 'contributors?per_page=1');
+    const cached = await redis.get(`github:info-${repo}`);
+    if (cached) {
+      return JSON.parse(cached);
+    }
 
-    if (res.status >= 400) throw new Error(`[Error ${res.status}] - ${(await res.json()).message}`);
+    const [statsParticipation, statsCodeFrequency, commitsInfo, contributorsInfo] =
+      await Promise.all([
+        getStatsParticipation(owner, repo),
+        getStatsCodeFrequency(owner, repo),
+        getCommitsInfo(owner, repo),
+        getContributorsInfo(owner, repo),
+      ]);
 
-    const contributorsResponse = await res.headers.get('link');
-    [, returnedValue.totalContributors] = contributorsResponse.match(
-      /.*"next".*&page=([0-9]*).*"last".*/
-    );
-  } catch (error) {
-    console.error(error);
-  }
+    returnGithubInfo = {
```

Comment:
You call this function, `getGitHubData` so let's call this `githubData` or just `data`.  An object shouldn't have a verb as part of its name (e.g., no `return`).  This is `data` and we'll `return` it later.

Reply:
Changed